### PR TITLE
missed wakeup when growing kmem cache

### PR DESCRIPTION
**Pull request by ahrens**
### Motivation and Context

When growing the size of a (VMEM or KVMEM) kmem cache, spl_cache_grow()
always does taskq_dispatch(spl_cache_grow_work), and then waits for the
KMC_BIT_GROWING to be cleared by the taskq thread.

The taskq thread (spl_cache_grow_work()) does:
1. allocate new slab and add to list
2. wake_up_all(skc_waitq)
3. clear_bit(KMC_BIT_GROWING)

Therefore, the waiting thread can wake up before GROWING has been
cleared.  It will see that the growing has not yet completed, and go
back to sleep until it hits the 100ms timeout.

This can have an extreme performance impact on workloads that alloc/free
more than fits in the (statically-sized) magazines.  These workloads
allocate and free slabs with high frequency.

The problem can be observed with `funclatency spl_cache_grow`, which on
some workloads shows that 99.5% of calls take <64us,  but we spend ~70% 
of our time in outliers, waiting for the 100ms timeout.

```
sudo /usr/share/bcc/tools/funclatency -d 10 spl_cache_grow
Tracing 1 functions for "spl_cache_grow"... Hit Ctrl-C to end.
     nsecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 0        |                                        |
       128 -> 255        : 0        |                                        |
       256 -> 511        : 0        |                                        |
       512 -> 1023       : 0        |                                        |
      1024 -> 2047       : 0        |                                        |
      2048 -> 4095       : 0        |                                        |
      4096 -> 8191       : 0        |                                        |
      8192 -> 16383      : 812      |**                                      |
     16384 -> 32767      : 13961    |****************************************|
     32768 -> 65535      : 2666     |*******                                 |
     65536 -> 131071     : 304      |                                        |
    131072 -> 262143     : 149      |                                        |
    262144 -> 524287     : 124      |                                        |
    524288 -> 1048575    : 106      |                                        |
   1048576 -> 2097151    : 36       |                                        |
   2097152 -> 4194303    : 13       |                                        |
   4194304 -> 8388607    : 1        |                                        |
   8388608 -> 16777215   : 0        |                                        |
  16777216 -> 33554431   : 0        |                                        |
  33554432 -> 67108863   : 0        |                                        |
  67108864 -> 134217727  : 69       |                                        |
```

### Description

The fix is to do `clear_bit(KMC_BIT_GROWING)` before
`wake_up_all(skc_waitq)`.

A future investigation should evaluate if we still actually need to
taskq_dispatch() at all, and if so on which kernel versions.

### How Has This Been Tested?

The more change more than doubles the performance of `zfs send` (with recordsize>=32K, otherwise we use the SLUB caches) (and with a bunch of other forthcoming performance enhancements to `zfs send`).

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/os/linux/spl/spl-kmem-cache.c [hunk 1]
```diff
@@ -1185,7 +1185,6 @@ __spl_cache_grow(spl_kmem_cache_t *skc, int flags)
 		smp_mb__before_atomic();
 		clear_bit(KMC_BIT_DEADLOCKED, &skc->skc_flags);
 		smp_mb__after_atomic();
-		wake_up_all(&skc->skc_waitq);
```
Comment (pcd1193182) [inline module/os/linux/spl/spl-kmem-cache.c]: Is there any potential race associated with doing the wakeup after we've released the spinlock? Otherwise this change seems good to me.

Comment (ahrens) [reply module/os/linux/spl/spl-kmem-cache.c]: I would think not, because we (obviously) don't hold the spin lock while waiting, it doesn't seem like we could use it to synchronize.  I suspect that there may be other more subtle race conditions inherent to wait_event/wake_up (e.g. we're about to go to sleep, but the wake_up() already happened), but it's a bit hard to follow the twisted web of macros and assembly.  Maybe @behlendorf can enlighten us.

Comment (behlendorf) [reply module/os/linux/spl/spl-kmem-cache.c]: Waking the waiters outside the spin lock is fine.  The `wake_up_all` was located there so that we only woke waiters when successfully allocating a new slab.  While it's unlikely this will fail it's definitely possible.  How about further updating `spl_cache_grow_work` like this to preserve this wake up behavior.

```diff
diff --git a/module/os/linux/spl/spl-kmem-cache.c b/module/os/linux/spl/spl-kmem-cache.c
index 7dd8e85..23477b0 100644
--- a/module/os/linux/spl/spl-kmem-cache.c
+++ b/module/os/linux/spl/spl-kmem-cache.c
@@ -1198,13 +1198,16 @@ spl_cache_grow_work(void *data)
        spl_kmem_alloc_t *ska = (spl_kmem_alloc_t *)data;
        spl_kmem_cache_t *skc = ska->ska_cache;
 
-       (void) __spl_cache_grow(skc, ska->ska_flags);
+       int error = __spl_cache_grow(skc, ska->ska_flags);
 
        atomic_dec(&skc->skc_ref);
        smp_mb__before_atomic();
        clear_bit(KMC_BIT_GROWING, &skc->skc_flags);
        smp_mb__after_atomic();
 
+       if (error == 0)
+               wake_up_all(&skc->skc_waitq);
+
        kfree(ska);
 }
```

Comment (ahrens) [reply module/os/linux/spl/spl-kmem-cache.c]: sure, done.
